main/lsp/requests/hover.cc

```diff
@@ -89,9 +89,15 @@ LSPResult LSPLoop::handleTextDocumentHover(unique_ptr<core::GlobalState> gs, con
             if (constraint) {
                 retType = core::Types::instantiate(core::Context(*gs, core::Symbols::root()), retType, *constraint);
             }
-            auto methodInfo = methodInfoString(*gs, retType, *sendResp->dispatchResult, constraint);
-            response->result =
-                make_unique<Hover>(formatHoverText(config.clientHoverMarkupKind, methodInfo, documentation));
+            if (sendResp->dispatchResult->main.method.exists() && sendResp->dispatchResult->main.method.isSynthetic()) {
+                // For synthetic methods, limit hover to the return type
+                response->result = make_unique<Hover>(
+                    formatHoverText(config.clientHoverMarkupKind, retType->showWithMoreInfo(*gs), documentation));
+            } else {
+                auto methodInfo = methodInfoString(*gs, retType, *sendResp->dispatchResult, constraint);
+                response->result =
+                    make_unique<Hover>(formatHoverText(config.clientHoverMarkupKind, methodInfo, documentation));
+            }
```

Comment:
non-blocking nit: seems like the only thing different between these two branches is the second argument to formatHoverText. I think this could be simplified by factoring that out after the if, and just computing a string in each branch of the `if` like

```cpp
string hoverText;
if (...) {
  hoverText = ...
} else {
  hoverText = ...
}

response->result = ...(..., hoverText, ...);
```

Reply:
Yup can do.